File: sdk/extension-sdk-ts/src/handlers.ts
import type {
  AgentOnSessionUpdatePayload,
  AgentPostSessionCreatePayload,
  AgentPostSessionEndPayload,
  AgentPreSessionCreatePayload,
  AgentPreSessionResumePayload,
  ArtifactPostWritePayload,
  ArtifactPreWritePayload,
  ArtifactWritePatch,
  BatchParamsPatch,
  EntriesPatch,
  Event,
  ExecuteHookRequest,
  ExtraSourcesPatch,
  FetchConfigPatch,
  GroupsPatch,
  HealthCheckRequest,
  HealthCheckResponse,
  HookContext,
  HookInfo,
  HookName,
  IssuesPatch,
  JobPatch,
  JobPostExecutePayload,
  JobPreExecutePayload,
  JobPreRetryPayload,
  JobsPatch,
  PlanPostDiscoverPayload,
  PlanPostGroupPayload,
  PlanPostPrepareJobsPayload,
  PlanPreDiscoverPayload,
  PlanPreGroupPayload,
  PlanPrePrepareJobsPayload,
  PromptPostBuildPayload,
  PromptPreBuildPayload,
  PromptPreSystemPayload,
  PromptTextPatch,
  ResolveDecisionPatch,
  ResumeSessionRequestPatch,
  ReviewPostFetchPayload,
  ReviewPostFixPayload,
  ReviewPreBatchPayload,
  ReviewPreFetchPayload,
  ReviewPreResolvePayload,
  ReviewWatchFinishedPayload,
  ReviewWatchPostRoundPayload,
  ReviewWatchPrePushPatch,
  ReviewWatchPrePushPayload,
  ReviewWatchPreRoundPatch,
  ReviewWatchPreRoundPayload,
  RetryDecisionPatch,
  RunPostShutdownPayload,
  RunPostStartPayload,
  RunPreShutdownPayload,
  RunPreStartPayload,
  RuntimeConfigPatch,
  SessionRequestPatch,
  ShutdownRequest,
  SystemAddendumPatch,
} from "./types.js";

/** Callback invoked when the host forwards a rc event to the extension. */
export type EventHandler = (event: Event) => Promise<void> | void;
/** Low-level hook handler that receives the raw JSON payload and returns a raw patch. */
export type RawHookHandler = (context: HookContext, payload: unknown) => Promise<unknown> | unknown;
/** Callback that overrides the default health check response. */
export type HealthCheckHandler = (
  request: HealthCheckRequest
) => Promise<HealthCheckResponse> | HealthCheckResponse;
/** Callback invoked during graceful shutdown before the process exits. */
export type ShutdownHandler = (request: ShutdownRequest) => Promise<void> | void;

/** Typed handler for mutable hooks. Receives a strongly-typed payload and returns an optional patch describing what to mutate. */
export type MutableHookHandler<Payload, Patch> = (
  context: HookContext,
  payload: Payload
) => Promise<Patch | void> | Patch | void;

/** Typed handler for observer-only hooks. Receives the payload for side-effects with no return value. */
export type ObserverHookHandler<Payload> = (
  context: HookContext,
  payload: Payload
) => Promise<void> | void;

/** Internal representation of a registered hook handler with its mutability flag. */
export interface RegisteredHook {
  mutable: boolean;
  handler: RawHookHandler;
}

/** Minimal interface for registering raw hook handlers on an extension. */
export interface HookRegistrationSurface {
  handle(hook: HookName, handler: RawHookHandler): this;
}

/** Type-safe mapping from each hook name to its corresponding handler signature, ensuring payload and patch types are correct at compile time. */
export type HookHandlerMatrix = {
  "plan.pre_discover": MutableHookHandler<PlanPreDiscoverPayload, ExtraSourcesPatch>;
  "plan.post_discover": MutableHookHandler<PlanPostDiscoverPayload, EntriesPatch>;
  "plan.pre_group": MutableHookHandler<PlanPreGroupPayload, EntriesPatch>;
  "plan.post_group": MutableHookHandler<PlanPostGroupPayload, GroupsPatch>;
  "plan.pre_prepare_jobs": MutableHookHandler<PlanPrePrepareJobsPayload, GroupsPatch>;
  "plan.post_prepare_jobs": MutableHookHandler<PlanPostPrepareJobsPayload, JobsPatch>;
  "prompt.pre_build": MutableHookHandler<PromptPreBuildPayload, BatchParamsPatch>;
  "prompt.post_build": MutableHookHandler<PromptPostBuildPayload, PromptTextPatch>;
  "prompt.pre_system": MutableHookHandler<PromptPreSystemPayload, SystemAddendumPatch>;
  "agent.pre_session_create": MutableHookHandler<AgentPreSessionCreatePayload, SessionRequestPatch>;
  "agent.post_session_create": ObserverHookHandler<AgentPostSessionCreatePayload>;
  "agent.pre_session_resume": MutableHookHandler<
    AgentPreSessionResumePayload,
    ResumeSessionRequestPatch
  >;
  "agent.on_session_update": ObserverHookHandler<AgentOnSessionUpdatePayload>;
  "agent.post_session_end": ObserverHookHandler<AgentPostSessionEndPayload>;
  "job.pre_execute": MutableHookHandler<JobPreExecutePayload, JobPatch>;
  "job.post_execute": ObserverHookHandler<JobPostExecutePayload>;
  "job.pre_retry": MutableHookHandler<JobPreRetryPayload, RetryDecisionPatch>;
  "run.pre_start": MutableHookHandler<RunPreStartPayload, RuntimeConfigPatch>;
  "run.post_start": ObserverHookHandler<RunPostStartPayload>;
  "run.pre_shutdown": ObserverHookHandler<RunPreShutdownPayload>;
  "run.post_shutdown": ObserverHookHandler<RunPostShutdownPayload>;
  "review.pre_fetch": MutableHookHandler<ReviewPreFetchPayload, FetchConfigPatch>;
  "review.post_fetch": MutableHookHandler<ReviewPostFetchPayload, IssuesPatch>;
  "review.pre_batch": MutableHookHandler<ReviewPreBatchPayload, GroupsPatch>;
  "review.post_fix": ObserverHookHandler<ReviewPostFixPayload>;
  "review.pre_resolve": MutableHookHandler<ReviewPreResolvePayload, ResolveDecisionPatch>;
  "review.watch_pre_round": MutableHookHandler<
    ReviewWatchPreRoundPayload,
    ReviewWatchPreRoundPatch
  >;
  "review.watch_post_round": ObserverHookHandler<ReviewWatchPostRoundPayload>;
  "review.watch_pre_push": MutableHookHandler<ReviewWatchPrePushPayload, ReviewWatchPrePushPatch>;
  "review.watch_finished": ObserverHookHandler<ReviewWatchFinishedPayload>;
  "artifact.pre_write": MutableHookHandler<ArtifactPreWritePayload, ArtifactWritePatch>;
  "artifact.post_write": ObserverHookHandler<ArtifactPostWritePayload>;
};

/** Returns true if the given hook name corresponds to a mutable hook that can return a patch. */
export function isMutableHook(hook: HookName): boolean {
  switch (hook) {
    case "agent.post_session_create":
    case "agent.on_session_update":
    case "agent.post_session_end":
    case "job.post_execute":
    case "run.post_start":
    case "run.pre_shutdown":
    case "run.post_shutdown":
    case "review.post_fix":
    case "review.watch_post_round":
    case "review.watch_finished":
    case "artifact.post_write":
      return false;
    default:
      return true;
  }
}

/** Registers a typed mutable hook handler on the given surface, wrapping it as a raw handler. */
export function registerMutableHook<Payload, Patch>(
  surface: HookRegistrationSurface,
  hook: HookName,
  handler: MutableHookHandler<Payload, Patch>
): HookRegistrationSurface {
  return surface.handle(hook, async (context, payload) => {
    return await handler(context, payload as Payload);
  });
}

/** Registers a typed observer hook handler on the given surface, wrapping it as a raw handler that returns undefined. */
export function registerObserverHook<Payload>(
  surface: HookRegistrationSurface,
  hook: HookName,
  handler: ObserverHookHandler<Payload>
): HookRegistrationSurface {
  return surface.handle(hook, async (context, payload) => {
    await handler(context, payload as Payload);
    return undefined;
  });
}

/** Builds a {@link HookContext} from an incoming execute-hook request and host metadata. */
export function requestContext(
  request: ExecuteHookRequest,
  host: HookContext["host"]
): HookContext {
  return {
    invocation_id: request.invocation_id,
    hook: request.hook as HookInfo,
    host,
  };
}

File: sdk/extension/handlers.go
package extension

import (
	"context"
	"encoding/json"
	"fmt"
)

func registerMutableHook[Payload any, Patch any](
	e *Extension,
	event HookName,
	handler func(context.Context, HookContext, Payload) (Patch, error),
) *Extension {
	if e == nil || handler == nil {
		return e
	}

	return e.Handle(event, func(ctx context.Context, hook HookContext, raw json.RawMessage) (json.RawMessage, error) {
		var payload Payload
		if err := unmarshalJSON(raw, &payload); err != nil {
			return nil, newInvalidParamsError(map[string]any{
				"method": "execute_hook",
				"hook":   string(event),
				"error":  err.Error(),
			})
		}

		patch, err := handler(ctx, hook, payload)
		if err != nil {
			return nil, err
		}
		return marshalJSON(patch)
	})
}

func registerObserverHook[Payload any](
	e *Extension,
	event HookName,
	handler func(context.Context, HookContext, Payload) error,
) *Extension {
	if e == nil || handler == nil {
		return e
	}

	return e.Handle(event, func(ctx context.Context, hook HookContext, raw json.RawMessage) (json.RawMessage, error) {
		var payload Payload
		if err := unmarshalJSON(raw, &payload); err != nil {
			return nil, newInvalidParamsError(map[string]any{
				"method": "execute_hook",
				"hook":   string(event),
				"error":  err.Error(),
			})
		}
		if err := handler(ctx, hook, payload); err != nil {
			return nil, err
		}
		return nil, nil
	})
}

// OnPlanPreDiscover registers the plan.pre_discover handler.
func (e *Extension) OnPlanPreDiscover(
	handler func(context.Context, HookContext, PlanPreDiscoverPayload) (ExtraSourcesPatch, error),
) *Extension {
	return registerMutableHook(e, HookPlanPreDiscover, handler)
}

// OnPlanPostDiscover registers the plan.post_discover handler.
func (e *Extension) OnPlanPostDiscover(
	handler func(context.Context, HookContext, PlanPostDiscoverPayload) (EntriesPatch, error),
) *Extension {
	return registerMutableHook(e, HookPlanPostDiscover, handler)
}

// OnPlanPreGroup registers the plan.pre_group handler.
func (e *Extension) OnPlanPreGroup(
	handler func(context.Context, HookContext, PlanPreGroupPayload) (EntriesPatch, error),
) *Extension {
	return registerMutableHook(e, HookPlanPreGroup, handler)
}

// OnPlanPostGroup registers the plan.post_group handler.
func (e *Extension) OnPlanPostGroup(
	handler func(context.Context, HookContext, PlanPostGroupPayload) (GroupsPatch, error),
) *Extension {
	return registerMutableHook(e, HookPlanPostGroup, handler)
}

// OnPlanPrePrepareJobs registers the plan.pre_prepare_jobs handler.
func (e *Extension) OnPlanPrePrepareJobs(
	handler func(context.Context, HookContext, PlanPrePrepareJobsPayload) (GroupsPatch, error),
) *Extension {
	return registerMutableHook(e, HookPlanPrePrepareJobs, handler)
}

// OnPlanPreResolveTaskRuntime registers the plan.pre_resolve_task_runtime handler.
func (e *Extension) OnPlanPreResolveTaskRuntime(
	handler func(context.Context, HookContext, PlanPreResolveTaskRuntimePayload) (TaskRuntimePatch, error),
) *Extension {
	return registerMutableHook(e, HookPlanPreResolveTaskRuntime, handler)
}

// OnPlanPostPrepareJobs registers the plan.post_prepare_jobs handler.
func (e *Extension) OnPlanPostPrepareJobs(
	handler func(context.Context, HookContext, PlanPostPrepareJobsPayload) (JobsPatch, error),
) *Extension {
	return registerMutableHook(e, HookPlanPostPrepareJobs, handler)
}

// OnPromptPreBuild registers the prompt.pre_build handler.
func (e *Extension) OnPromptPreBuild(
	handler func(context.Context, HookContext, PromptPreBuildPayload) (BatchParamsPatch, error),
) *Extension {
	return registerMutableHook(e, HookPromptPreBuild, handler)
}

// OnPromptPostBuild registers the prompt.post_build handler.
func (e *Extension) OnPromptPostBuild(
	handler func(context.Context, HookContext, PromptPostBuildPayload) (PromptTextPatch, error),
) *Extension {
	return registerMutableHook(e, HookPromptPostBuild, handler)
}

// OnPromptPreSystem registers the prompt.pre_system handler.
func (e *Extension) OnPromptPreSystem(
	handler func(context.Context, HookContext, PromptPreSystemPayload) (SystemAddendumPatch, error),
) *Extension {
	return registerMutableHook(e, HookPromptPreSystem, handler)
}

// OnAgentPreSessionCreate registers the agent.pre_session_create handler.
func (e *Extension) OnAgentPreSessionCreate(
	handler func(context.Context, HookContext, AgentPreSessionCreatePayload) (SessionRequestPatch, error),
) *Extension {
	return registerMutableHook(e, HookAgentPreSessionCreate, handler)
}

// OnAgentPostSessionCreate registers the agent.post_session_create handler.
func (e *Extension) OnAgentPostSessionCreate(
	handler func(context.Context, HookContext, AgentPostSessionCreatePayload) error,
) *Extension {
	return registerObserverHook(e, HookAgentPostSessionCreate, handler)
}

// OnAgentPreSessionResume registers the agent.pre_session_resume handler.
func (e *Extension) OnAgentPreSessionResume(
	handler func(context.Context, HookContext, AgentPreSessionResumePayload) (ResumeSessionRequestPatch, error),
) *Extension {
	return registerMutableHook(e, HookAgentPreSessionResume, handler)
}

// OnAgentOnSessionUpdate registers the agent.on_session_update handler.
func (e *Extension) OnAgentOnSessionUpdate(
	handler func(context.Context, HookContext, AgentOnSessionUpdatePayload) error,
) *Extension {
	return registerObserverHook(e, HookAgentOnSessionUpdate, handler)
}

// OnAgentPostSessionEnd registers the agent.post_session_end handler.
func (e *Extension) OnAgentPostSessionEnd(
	handler func(context.Context, HookContext, AgentPostSessionEndPayload) error,
) *Extension {
	return registerObserverHook(e, HookAgentPostSessionEnd, handler)
}

// OnJobPreExecute registers the job.pre_execute handler.
func (e *Extension) OnJobPreExecute(
	handler func(context.Context, HookContext, JobPreExecutePayload) (JobPatch, error),
) *Extension {
	return registerMutableHook(e, HookJobPreExecute, handler)
}

// OnJobPostExecute registers the job.post_execute handler.
func (e *Extension) OnJobPostExecute(
	handler func(context.Context, HookContext, JobPostExecutePayload) error,
) *Extension {
	return registerObserverHook(e, HookJobPostExecute, handler)
}

// OnJobPreRetry registers the job.pre_retry handler.
func (e *Extension) OnJobPreRetry(
	handler func(context.Context, HookContext, JobPreRetryPayload) (RetryDecisionPatch, error),
) *Extension {
	return registerMutableHook(e, HookJobPreRetry, handler)
}

// OnRunPreStart registers the run.pre_start handler.
func (e *Extension) OnRunPreStart(
	handler func(context.Context, HookContext, RunPreStartPayload) (RuntimeConfigPatch, error),
) *Extension {
	return registerMutableHook(e, HookRunPreStart, handler)
}

// OnRunPostStart registers the run.post_start handler.
func (e *Extension) OnRunPostStart(
	handler func(context.Context, HookContext, RunPostStartPayload) error,
) *Extension {
	return registerObserverHook(e, HookRunPostStart, handler)
}

// OnRunPreShutdown registers the run.pre_shutdown handler.
func (e *Extension) OnRunPreShutdown(
	handler func(context.Context, HookContext, RunPreShutdownPayload) error,
) *Extension {
	return registerObserverHook(e, HookRunPreShutdown, handler)
}

// OnRunPostShutdown registers the run.post_shutdown handler.
func (e *Extension) OnRunPostShutdown(
	handler func(context.Context, HookContext, RunPostShutdownPayload) error,
) *Extension {
	return registerObserverHook(e, HookRunPostShutdown, handler)
}

// OnReviewPreFetch registers the review.pre_fetch handler.
func (e *Extension) OnReviewPreFetch(
	handler func(context.Context, HookContext, ReviewPreFetchPayload) (FetchConfigPatch, error),
) *Extension {
	return registerMutableHook(e, HookReviewPreFetch, handler)
}

// OnReviewPostFetch registers the review.post_fetch handler.
func (e *Extension) OnReviewPostFetch(
	handler func(context.Context, HookContext, ReviewPostFetchPayload) (IssuesPatch, error),
) *Extension {
	return registerMutableHook(e, HookReviewPostFetch, handler)
}

// OnReviewPreBatch registers the review.pre_batch handler.
func (e *Extension) OnReviewPreBatch(
	handler func(context.Context, HookContext, ReviewPreBatchPayload) (GroupsPatch, error),
) *Extension {
	return registerMutableHook(e, HookReviewPreBatch, handler)
}

// OnReviewPostFix registers the review.post_fix handler.
func (e *Extension) OnReviewPostFix(
	handler func(context.Context, HookContext, ReviewPostFixPayload) error,
) *Extension {
	return registerObserverHook(e, HookReviewPostFix, handler)
}

// OnReviewPreResolve registers the review.pre_resolve handler.
func (e *Extension) OnReviewPreResolve(
	handler func(context.Context, HookContext, ReviewPreResolvePayload) (ResolveDecisionPatch, error),
) *Extension {
	return registerMutableHook(e, HookReviewPreResolve, handler)
}

// OnReviewWatchPreRound registers the review.watch_pre_round handler.
func (e *Extension) OnReviewWatchPreRound(
	handler func(context.Context, HookContext, ReviewWatchPreRoundPayload) (ReviewWatchPreRoundPatch, error),
) *Extension {
	return registerMutableHook(e, HookReviewWatchPreRound, handler)
}

// OnReviewWatchPostRound registers the review.watch_post_round handler.
func (e *Extension) OnReviewWatchPostRound(
	handler func(context.Context, HookContext, ReviewWatchPostRoundPayload) error,
) *Extension {
	return registerObserverHook(e, HookReviewWatchPostRound, handler)
}

// OnReviewWatchPrePush registers the review.watch_pre_push handler.
func (e *Extension) OnReviewWatchPrePush(
	handler func(context.Context, HookContext, ReviewWatchPrePushPayload) (ReviewWatchPrePushPatch, error),
) *Extension {
	return registerMutableHook(e, HookReviewWatchPrePush, handler)
}

// OnReviewWatchFinished registers the review.watch_finished handler.
func (e *Extension) OnReviewWatchFinished(
	handler func(context.Context, HookContext, ReviewWatchFinishedPayload) error,
) *Extension {
	return registerObserverHook(e, HookReviewWatchFinished, handler)
}

// OnArtifactPreWrite registers the artifact.pre_write handler.
func (e *Extension) OnArtifactPreWrite(
	handler func(context.Context, HookContext, ArtifactPreWritePayload) (ArtifactWritePatch, error),
) *Extension {
	return registerMutableHook(e, HookArtifactPreWrite, handler)
}

// OnArtifactPostWrite registers the artifact.post_write handler.
func (e *Extension) OnArtifactPostWrite(
	handler func(context.Context, HookContext, ArtifactPostWritePayload) error,
) *Extension {
	return registerObserverHook(e, HookArtifactPostWrite, handler)
}

// MustJSON marshals value into json.RawMessage or panics.
func MustJSON(value any) json.RawMessage {
	raw, err := marshalJSON(value)
	if err != nil {
		panic(fmt.Sprintf("marshal json: %v", err))
	}
	return raw
}
